File: pyorion/api/dirs.py
# Copyright 2025-2030 Ari Bermeki @ YellowSiC within The Commons Conservancy
# SPDX-License-Identifier: Apache-2.0
# SPDX-License-Identifier: MIT

"""Dirs API - Access to system standard directories.

Wrapper around the Rust ``dirs`` crate via the PyOrion event loop.
"""

from typing import Optional

from pyorion.runtime.runtime_handle import event_register


class DirsAPI:
    """Asynchronous API wrapper for system standard directories.

    Provides access to common user and system paths such as home,
    cache, config, documents, downloads, etc.
    """

    def __init__(self) -> None:
        """Initialize a new :class:`DirsAPI` instance."""

    async def home_dir(self) -> Optional[str]:
        """Get the path to the user's home directory.

        :return: Absolute path to the home directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.homeDir")

    async def cache_dir(self) -> Optional[str]:
        """Get the path to the cache directory.

        :return: Absolute path to the cache directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.cacheDir")

    async def config_dir(self) -> Optional[str]:
        """Get the path to the global configuration directory.

        :return: Absolute path to the config directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.configDir")

    async def config_local_dir(self) -> Optional[str]:
        """Get the path to the user-specific configuration directory.

        :return: Absolute path to the local config directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.configLocalDir")

    async def data_dir(self) -> Optional[str]:
        """Get the path to the global data directory.

        :return: Absolute path to the data directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.dataDir")

    async def data_local_dir(self) -> Optional[str]:
        """Get the path to the user-specific data directory.

        :return: Absolute path to the local data directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.dataLocalDir")

    async def desktop_dir(self) -> Optional[str]:
        """Get the path to the desktop directory.

        :return: Absolute path to the desktop directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.desktopDir")

    async def document_dir(self) -> Optional[str]:
        """Get the path to the documents directory.

        :return: Absolute path to the documents directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.documentDir")

    async def download_dir(self) -> Optional[str]:
        """Get the path to the downloads directory.

        :return: Absolute path to the downloads directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.downloadDir")

    async def executable_dir(self) -> Optional[str]:
        """Get the path to the executables directory.

        :return: Absolute path to the executables directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.executableDir")

    async def font_dir(self) -> Optional[str]:
        """Get the path to the system fonts directory.

        :return: Absolute path to the fonts directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.fontDir")

    async def picture_dir(self) -> Optional[str]:
        """Get the path to the pictures directory.

        :return: Absolute path to the pictures directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.pictureDir")

    async def preference_dir(self) -> Optional[str]:
        """Get the path to the preferences directory.

        :return: Absolute path to the preferences directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.preferenceDir")

    async def public_dir(self) -> Optional[str]:
        """Get the path to the public (shared) directory.

        :return: Absolute path to the public directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.publicDir")

    async def runtime_dir(self) -> Optional[str]:
        """Get the path to the runtime directory.

        :return: Absolute path to the runtime directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.runtimeDir")

    async def state_dir(self) -> Optional[str]:
        """Get the path to the state directory.

        :return: Absolute path to the state directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.stateDir")

    async def template_dir(self) -> Optional[str]:
        """Get the path to the templates directory.

        :return: Absolute path to the templates directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.templateDir")

    async def video_dir(self) -> Optional[str]:
        """Get the path to the videos directory.

        :return: Absolute path to the videos directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.videoDir")

    async def audio_dir(self) -> Optional[str]:
        """Get the path to the audio/music directory.

        :return: Absolute path to the audio directory or ``None`` if unavailable.
        :rtype: Optional[str]
        """
        return await event_register("dirs.audioDir")

File: src/api/dirs.rs
// Copyright 2025-2030 Ari Bermeki @ YellowSiC within The Commons Conservancy
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

use crate::api_manager::ApiManager;
use anyhow::Result;
use pyorion_macros::api;
use serde_json::json;

pub fn dirs_api(api_manager: &mut ApiManager) {
    api_manager.register_api("dirs.homeDir", home_dir);
    api_manager.register_api("dirs.cacheDir", cache_dir);
    api_manager.register_api("dirs.configDir", config_dir);
    api_manager.register_api("dirs.configLocalDir", config_local_dir);
    api_manager.register_api("dirs.dataDir", data_dir);
    api_manager.register_api("dirs.dataLocalDir", data_local_dir);
    api_manager.register_api("dirs.desktopDir", desktop_dir);
    api_manager.register_api("dirs.documentDir", document_dir);
    api_manager.register_api("dirs.downloadDir", download_dir);
    api_manager.register_api("dirs.executableDir", executable_dir);
    api_manager.register_api("dirs.fontDir", font_dir);
    api_manager.register_api("dirs.pictureDir", picture_dir);
    api_manager.register_api("dirs.preferenceDir", preference_dir);
    api_manager.register_api("dirs.publicDir", public_dir);
    api_manager.register_api("dirs.runtimeDir", runtime_dir);
    api_manager.register_api("dirs.stateDir", state_dir);
    api_manager.register_api("dirs.templateDir", template_dir);
    api_manager.register_api("dirs.videoDir", video_dir);
    api_manager.register_api("dirs.audioDir", audio_dir);
}

#[api]
fn home_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::home_dir()))
}

#[api]
fn cache_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::cache_dir()))
}
#[api]
pub fn file_name(path: String) -> Result<Option<String>> {
    Ok(std::path::Path::new(&path)
        .file_name()
        .map(|name| name.to_string_lossy().into_owned()))
}

#[api]
fn config_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::config_dir()))
}

#[api]
fn config_local_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::config_local_dir()))
}

#[api]
fn data_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::data_dir()))
}

#[api]
fn data_local_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::data_local_dir()))
}

#[api]
fn desktop_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::desktop_dir()))
}

#[api]
fn document_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::document_dir()))
}

#[api]
fn download_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::download_dir()))
}

#[api]
fn executable_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::executable_dir()))
}

#[api]
fn font_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::font_dir()))
}

#[api]
fn picture_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::picture_dir()))
}

#[api]
fn preference_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::preference_dir()))
}

#[api]
fn public_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::public_dir()))
}

#[api]
fn runtime_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::runtime_dir()))
}

#[api]
fn state_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::state_dir()))
}

#[api]
fn template_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::template_dir()))
}

#[api]
fn video_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::video_dir()))
}

#[api]
fn audio_dir() -> Result<serde_json::Value> {
    Ok(json!(dirs::audio_dir()))
}
